vectorstores/pgvector/pgvector.go

```diff
@@ -0,0 +1,329 @@
+package pgvector
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/google/uuid"
+	"github.com/jackc/pgx/v5"
+	"github.com/pgvector/pgvector-go"
+	"github.com/tmc/langchaingo/embeddings"
+	"github.com/tmc/langchaingo/schema"
+	"github.com/tmc/langchaingo/vectorstores"
+)
+
+var (
+	ErrEmbedderWrongNumberVectors = errors.New("number of vectors from embedder does not match number of documents")
+	ErrInvalidScoreThreshold      = errors.New("score threshold must be between 0 and 1")
+	ErrInvalidFilters             = errors.New("invalid filters")
+	ErrUnsupportedOptions         = errors.New("unsupported options")
+)
+
+// Store is a wrapper around the pgvector client.
+type Store struct {
+	embedder              embeddings.Embedder
+	conn                  *pgx.Conn
+	postgresConnectionURL string
+	embeddingTableName    string
+	collectionTableName   string
+	collectionName        string
+	collectionUUID        string
+	collectionMetadata    map[string]any
+	preDeleteCollection   bool
+}
+
+var _ vectorstores.VectorStore = Store{}
+
+// New creates a new Store with options.
+func New(ctx context.Context, opts ...Option) (Store, error) {
+	store, err := applyClientOptions(opts...)
+	if err != nil {
+		return Store{}, err
+	}
+	store.conn, err = pgx.Connect(ctx, store.postgresConnectionURL)
+	if err != nil {
+		return Store{}, err
+	}
+
+	if err = store.conn.Ping(ctx); err != nil {
+		return Store{}, err
+	}
+
+	if err = store.createVectorExtensionIfNotExists(ctx); err != nil {
+		return Store{}, err
+	}
+	if err = store.createCollectionTableIfNotExists(ctx); err != nil {
+		return Store{}, err
+	}
+	if err = store.createEmbeddingTableIfNotExists(ctx); err != nil {
+		return Store{}, err
+	}
+	if store.preDeleteCollection {
+		if err = store.RemoveCollection(ctx); err != nil {
+			return Store{}, err
+		}
+	}
+	if store.collectionUUID, err = store.createOrGetCollection(ctx); err != nil {
+		return Store{}, err
+	}
+	return store, nil
+}
+
+func (s Store) createVectorExtensionIfNotExists(ctx context.Context) error {
+	tx, err := s.conn.Begin(ctx)
+	if err != nil {
+		return err
+	}
+	// inspired by
+	// https://github.com/langchain-ai/langchain/blob/v0.0.340/libs/langchain/langchain/vectorstores/pgvector.py#L167
+	// The advisor lock fixes issue arising from concurrent
+	// creation of the vector extension.
+	// https://github.com/langchain-ai/langchain/issues/12933
+	// For more information see:
+	// https://www.postgresql.org/docs/16/explicit-locking.html#ADVISORY-LOCKS
+	if _, err := tx.Exec(ctx, "SELECT pg_advisory_xact_lock(1573678846307946495)"); err != nil {
```

Comment:
Even if it's taken from elsewhere, it'd be good to describe the magic number passed to `pg_advisory_xact_lock`

Reply:
IMO, these magic number just not overlap is enough, from https://www.postgresql.org/docs/current/functions-admin.html#FUNCTIONS-ADVISORY-LOCKS.

I used 1573678846307946495, 1573678846307946494, 1573678846307946493 just to distinguish it from [1573678846307946496 in langchain](https://github.com/langchain-ai/langchain/blob/v0.0.340/libs/langchain/langchain/vectorstores/pgvector.py#L174).

Reply:
OK, but still it'd be better to have these as package-level constants, something like `const pgLockId<purpose> = ....`